### ISSUE TITLE
Missing guidance for migration from classical mappings to declarative mappings

### ISSUE THREAD

**Issue by tlandschoff-scale** [sqlalchemy/sqlalchemy#4698]
Hello world,

we have a middle sized application that is using SQLAlchemy as the ORM. We started in 2011 with SQLAlchemy 0.6 when the declarative mapping style was still a new feature (and there was [Elixir](https://pypi.org/project/Elixir/) as an alternative).

Today we are still struggling with our mapping configuration. As our data model is mostly hierarchic (a version control system in a database) we have problems with circular dependencies.

With declarative this would be a non-issue given forward declarations. Also, IDE support would be much better (currently most column accesses are flagged as error given that the descriptors are added by calling the mapper on the existing class). Finally, to work around some issues we have classes that don't have any ORM related stuff in their home modules but a final `foo_bar_mapper.py` import attaches the mapper for multiple related classes.

This really sucks, therefore there have been multiple attempts to convert our code base to use the declarative extension. Some leaf classes have been converted but the core is still stuck.

We tried to go from the base classes downwards or from the leaf classes upward in conversion. But incremental conversion has been unsuccessful.

Therefore I would really like some guidance in the SQLAlchemy documentation about how to approach converting a non-simple code base from classical to declarative mapping. Are there any success stories to do this in an incremental way?

I'd like to help out here but so far all attempts were a dead end wrt. the hard parts.

Thanks and greetings,

Torsten

**Comment by zzzeek** [comment]
Hi Torsten -

a simple question first, when you request "some guidance in the SQLAlchemy documentation", is it more important that the official documentation cover this, or would you just like help with your problem?    I'm not sure that an official section in the main docs is that useful in general, as the documentation is already huge and so to that extent at least tries to be present and future oriented, with the exception of the migration notes sections that each focus on individual issues.     As I'm sure you're aware, the vast majority of applications are using declarative style mappings which are only becoming more of the default in future SQLAlchemy versions where we will finally be moving "declarative" out of being an "extension" and into the main ORM.    However note that the classical "mapper()" call is not actually going away, it will work differently but we have had users who request that they can still map in this way.   But in the big picture, migrations like the one you are undertaking are not a general use case and also not one that we expect to be common going forward.    

There can certainly be things like blog posts and stuff elsewhere that is dated material but I don't have a good insight on how such a section could be worth maintaining in the main docs.   Not the least of which is that I don't really know the nature of the problems you are having.    As you are aware, the declarative process is essentially a metaclass that scans a class definition, sets a bunch of things up, and then ultimately calls the exact same mapper() function that your classical mappings are already invoking.       

Given that, the only area I can imagine declarative interfering with a class-base that is only partially based on declarative is if you have deep inheritance hierarchies, and declarative is attempting to apply its own mapping towards subclasses that you intend to map classically.   The other way around, using declarative to extend from a classically mapped base, I don't foresee any issues with that however it's not a regular use case so perhaps something goes wrong there as well.  But the bigger picture with this area is that most large applications will have lots of independent class hierarchies that are related to each other with relationships but not through common mapped superclasses, so issues with declarative's metaclass moving up and down the chain shouldn't be that much of an issue.

So to sum up for starters, a. I don't think the main documentation is necessarily a good place for this but also b. I don't know what your actual problem is so I don't have a clear view on what I'd be documenting in any case.

If you can produce some specific examples of patterns that were "unsuccessful" and show me reproduction cases, that would be a good start!

**Comment by tlandschoff-scale** [comment]
> is it more important that the official documentation cover this, 

It just looks like an omission to me in contrast to the otherwise excellent documentation.

> or would you just like help with your problem?

Currently it looks like we don't need help with this anymore as I just tried to do the migration in one big leap and it currently looks like the result is working. I still fear there might be some regressions though but I'd rather push forward with this now.

So no, I guess we got a handle on it.

> As I'm sure you're aware, the vast majority of applications are using declarative style mappings which are only becoming more of the default in future

Absolutely. It also seems to scale much better with big schemas. Especially the syntax

```
address = relation("Address")
```

is very helpful in avoiding dependency cycles in Python imports.

> migrations like the one you are undertaking are not a general use case and also not one that we expect to be common going forward

In that case it makes sense to close this ticket. I had the hope that we are not alone in trying a migration like this and could benefit from the experience other people had with this task.

>  The other way around, using declarative to extend from a classically mapped base, I don't foresee any issues with that however it's not a regular use case so perhaps something goes wrong there as well.

That's an important information and is also the process we ended up doing: Convert anything to declarative starting at the leaf/concrete classes.  We were using mixin classes together with classical mappings (duplicating some column and relationship configuration) and we often add features in the mixed in functionality.

There the declarative features would have been really helpful and we tried with little success to introduce declarative in the middle of the class hierarchy. I also once started a conversion going from the base class of our central and oldest class tree.

> If you can produce some specific examples of patterns that were "unsuccessful" and show me reproduction cases, that would be a good start!

I did on the previous attempts like in #4321. This time I also hit a bug (I think) that I will report separately.

So, to conclude this: It seems that not many users are running into this so no need to add a section to the documentation.

Thanks for your reply!

Greetings, Torsten

**Comment by zzzeek** [comment]
OK...so....maybe open a new issue with specific things I can work on..thanks!

**Comment by tlandschoff-scale** [comment]
Given that I just spent a while to exercise myself in migrating a toy example from classical mapping to declarative mapping, I wonder if this could serve as a tutorial of such a migration.

Here are the steps that I did: https://github.com/tlandschoff-scale/sqlalchemy-gh4699/compare/original-classical-mapping...final-declarative-mapping

Maybe this can be of use to somebody attempting the same thing.
